### ISSUE TITLE
Fails from undefined parameters without asking to define parameters

### ISSUE THREAD

**Issue by dickfrazier** [timmcmic/DLConversionV2#10]
When trying to run I was able to install and import the module, it asks for some parameters but not all the ones I expect and then fails and closes the powershell window.  I am probably doing it wrong, where do I set the params ahead of time like I did with V1? The parameters it asks for and I give it before it fails are:

groupSMTPAddress
globalCatalogServer
activeDirectoryCredential
logFolderPath
dnNoSyncOU

![image](https://user-images.githubusercontent.com/89863919/134374798-376f21a1-63aa-48d5-852a-4384cca76692.png)

`[9/22/2021 9:02:38 AM] - ********************************************************************************
[9/22/2021 9:02:38 AM] - ENTERING PARAMTER VALIDATION
[9/22/2021 9:02:38 AM] - ********************************************************************************
[9/22/2021 9:02:38 AM] - Validating that both AADConnectServer and AADConnectCredential are specified
[9/22/2021 9:02:38 AM] - Neither AADConnect Server or AADConnect Credentials specified - retain useAADConnect FALSE - False
[9/22/2021 9:02:38 AM] - Validating that both ExchangeServer and ExchangeCredential are specified.
[9/22/2021 9:02:38 AM] - Neither Exchange Server or Exchange Credentials specified - retain useOnPremisesExchange FALSE - False
[9/22/2021 9:02:38 AM] - Validating Exchange Online Credentials.
[9/22/2021 9:02:38 AM] - ERROR:  One permissions method to connect to Exchange Online must be specified.
[9/22/2021 9:02:38 AM] - ********************************************************************************
[9/22/2021 9:02:38 AM] - BEGIN disable-allPowerShellSessions
[9/22/2021 9:02:38 AM] - ********************************************************************************
[9/22/2021 9:02:38 AM] - Gathering all PS Sessions
[9/22/2021 9:02:38 AM] - Disconnecting Exchange Online Session
[9/22/2021 9:02:38 AM] - END disable-allPowerShellSessions
[9/22/2021 9:02:38 AM] - ********************************************************************************
[9/22/2021 9:02:38 AM] - Archiving files associated with run.
[9/22/2021 9:02:38 AM] - Split string for group name.
C:
DLConversionV2
logs
DLMigration
#hashtagtestdltoconvert.log
[9/22/2021 9:02:38 AM] - Split string for group name.
#hashtagtestdltoconvert
log
[9/22/2021 9:02:38 AM] - FAILED - renaming directory.
[9/22/2021 9:02:38 AM] - 20210922T0902384613-#hashtagtestdltoconvert-FAILED
[9/22/2021 9:02:38 AM] - C:\DLConversionV2\logs\DLMigration\`

**Comment by timmcmic** [comment]
Start here.

https://timmcmic.wordpress.com/2021/04/25/4116/

Tim

**Comment by timmcmic** [comment]
@dickfrazier I went through and validated the mandatories.  Exchange Online credentials were missing - but the rest of the validation checks checked out.  I ran just the start command and filled in the requested attributes.

**Comment by timmcmic** [comment]
And actually I just corrected myself - they're intentionally not mandatory.  Instead of using parameter sets because it became highly confusing i just used based validation code.  For exchange online - there's two methods to authenticate.  As of now when running _without_specifying parameters it does not prompt you for everything you need.

**Comment by dickfrazier** [comment]
Thanks! I make some real novice mistakes sometimes.  I now understand that I need to set all the params after start-distributionlistmigration instead of being prompted for all of them or setting them in a file.  I'm testing now 👍

**Comment by timmcmic** [comment]
Not all of them - part 1 lists the majors.  They're kinda in groups.

So required -> mail, active directory, storage location, the OU to put them in, and one set of exchange online credentials (either app based or user name and password).

From there - it depends on what you want the script to do.  If you want to shorten the migration time by having the script trigger a delta sync -> then you have to specify the ad connect information.

If you want to enable hybrid mail flow - then you need the exchange servers...

ETC...

**Comment by dickfrazier** [comment]
Is this understanding correct?: If I am not worried about the hybrid mail flow then it will function like V1 but not have to talk to the on prem exchange server?

**Comment by dickfrazier** [comment]
Also, thank you so much for putting so much time and energy into this and then also helping me understand everything, it is really  generous

**Comment by timmcmic** [comment]
Correct - it has been re-written to be based off LDAP - for customers that no longer have Exchange <or> do not need to enable hybrid mail flow.  So you are correct - there's no reason to specify an exchange server if you do not require hybrid mail flow.  As for V1 - this is a fairly major re-write the outcomes are the same but the mechanics are different.

**Comment by timmcmic** [comment]
Looks like you have an email address on the object that is not an accepted domain in Office 365.

**Comment by dickfrazier** [comment]
It was the subdomain that we use for the internal domain that cause the issue, I had added it after seeing the error once but it hadn't seen the change yet, running it a third time was successful, as have the other 8 DL's that I have done so far. About 70 to go! Thanks so much for your work

**Comment by timmcmic** [comment]
Fantastic.  I think I may add a check for the accepted domains.  It's a good idea - first time I've run into that but it's not terribly uncommon.

**Comment by dickfrazier** [comment]
One more thing that I found was that when it failed in that state I had to do some cleanup to re-run the script for that DL

1. Delete the new DL that was created- there will now be a DL with just a  testdl@orgname.mail.onmicrosoft.com address that is cloud based that must be deleted first (otherwise will fail with DL already exists error on re-run)
2. Move the DL from the non-sync OU back to a normal OU
3. Run a delta sync
4. re-run script for the DL

![image](https://user-images.githubusercontent.com/89863919/134428397-0761882c-0584-49fc-900a-c4308b979e0d.png)
